nexus/reconfigurator/planning/src/planner.rs

```diff
@@ -1227,19 +1229,72 @@ impl<'a> Planner<'a> {
+            zone_id: OmicronZoneUuid,
+            zone_kind: ZoneKind,
+            reason: UpdatingReason<'a>,
+        }
+
+        #[derive(Debug)]
+        #[expect(dead_code)]
+        enum UpdatingReason<'a> {
+            ImageSourceMismatch {
+                bp_image_source: &'a BlueprintZoneImageSource,
+                inv_image_source: &'a OmicronZoneImageSource,
+            },
+            MissingInInventory {
+                bp_image_source: &'a BlueprintZoneImageSource,
+            },
+        }
+
         for &sled_id in &sleds {
+            // Build a list of zones currently in the blueprint but where
+            // inventory has a mismatch or does not know about the zone.
+            //
+            // What about the case where a zone is in inventory but not in the
+            // blueprint? See
+            // https://github.com/oxidecomputer/omicron/issues/8589.
             let zones_currently_updating = self
                 .blueprint
                 .current_sled_zones(
                     sled_id,
                     BlueprintZoneDisposition::is_in_service,
                 )
                 .filter_map(|zone| {
-                    let image_source = zone.image_source.clone().into();
-                    if inventory_zones.get(&zone.id) != Some(&image_source) {
-                        Some((zone.id, zone.zone_type.kind()))
-                    } else {
-                        None
+                    let bp_image_source =
+                        OmicronZoneImageSource::from(zone.image_source.clone());
+                    match inventory_zones.get(&zone.id) {
+                        Some(inv_image_source)
+                            if inv_image_source == &bp_image_source =>
+                        {
+                            // The inventory and blueprint image sources match
+                            // -- this means that the zone is up-to-date.
+                            None
+                        }
+                        Some(inv_image_source) => {
+                            // The inventory and blueprint image sources differ.
+                            Some(ZoneCurrentlyUpdating {
+                                zone_id: zone.id,
+                                zone_kind: zone.kind(),
+                                reason: UpdatingReason::ImageSourceMismatch {
+                                    bp_image_source: &zone.image_source,
+                                    inv_image_source,
+                                },
+                            })
+                        }
+                        None => {
+                            // The blueprint has a zone that inventory does not have.
+                            Some(ZoneCurrentlyUpdating {
+                                zone_id: zone.id,
+                                zone_kind: zone.kind(),
+                                reason: UpdatingReason::MissingInInventory {
```

Comment:
I think this is fine but I'll just point out in case you want to be more specific - the inventory map we're looking at here is produced by `all_running_omicron_zones()`, which [filters out zones that have errors](https://github.com/oxidecomputer/omicron/blob/248e7ed22b4123ad486fe28394b086e5032b7906/nexus-sled-agent-shared/src/inventory.rs#L159). So this case is really "not running", which might be "missing" or might be "failed to start".

Reply:
Done.